src/Compiler/Checking/import.fs

```diff
@@ -72,23 +77,38 @@ type [<Struct; NoComparison; CustomEquality>] TTypeCacheKey =
             elif this.ty1 === other.ty1 && this.ty2 === other.ty2 then
                 true
             else
-                stampEquals this.tcGlobals this.ty1 other.ty1
-                && stampEquals this.tcGlobals this.ty2 other.ty2
+                HashStamps.stampEquals this.ty1 other.ty1
+                && HashStamps.stampEquals this.ty2 other.ty2
 
     override this.Equals(other:objnull) =
         match other with
         | :? TTypeCacheKey as p -> (this :> System.IEquatable<TTypeCacheKey>).Equals p
         | _ -> false
 
     override this.GetHashCode() : int =
-        let g = this.tcGlobals
-
-        let ty1Hash = combineHash (hashStamp g this.ty1) (hashTType g this.ty1)
-        let ty2Hash = combineHash (hashStamp g this.ty2) (hashTType g this.ty2)
-
-        let combined = combineHash (combineHash ty1Hash ty2Hash) (hash this.canCoerce)
+        HashStamps.hashTType this.ty1
+        |> pipeToHash (HashStamps.hashTType this.ty2)
+        |> pipeToHash (hash this.canCoerce)
+
+    override this.ToString () = $"{this.ty1.DebugText}-{this.ty2.DebugText}"
+
+let createTypeSubsumptionCache (g: TcGlobals) =
+    let options =
+        match g.compilationMode with
+        | CompilationMode.OneOff ->
+            // This is a one-off compilation, so we don't need to worry about eviction.
+            { CacheOptions.Default with
+                MaximumCapacity = 4 * 1024
+                EvictionMethod = EvictionMethod.NoEviction }
+        | _ ->
+            // Incremental use, so we need to set up the cache with eviction.
+            { CacheOptions.Default with
+                EvictionMethod = EvictionMethod.Background
+                PercentageToEvict = 5
+                MaximumCapacity = 4 * 32768 }
+    Cache.Create<TTypeCacheKey, bool>(options)
 
-        combined
+let typeSubsumptionCaches = ConditionalWeakTable<TcGlobals, Cache<TTypeCacheKey, bool>>()
```

Comment:
How often are TcGlobals instances replaced under editing? Only when a significant change occurs, or all the time?

ConditionalWeakTable does not care about custom equality, it is always pointer/instance-attached

Reply:
Not often, but this is not good for another reason. TcGlobals are generously cached at various places, in IncrementalBuilder and TransparentCompiler too. Attaching to it means kept alive forever. What I'm thinking now is dogfooding squared and just repace this with Cache<TcGlobals, Cache<_, _>>.